File: app/src/main/java/com/example/turboautismdoselog/DrugAdapter.java
package com.example.turboautismdoselog;

import android.content.Intent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.recyclerview.widget.RecyclerView;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.List;
import java.util.Locale;

public class DrugAdapter extends RecyclerView.Adapter<DrugAdapter.ViewHolder> {

    private List<DrugEntry> entries;
    private OnItemLongClickListener listener;

    public interface OnItemLongClickListener {
        void onItemLongClick(DrugEntry entry);
    }

    public DrugAdapter(List<DrugEntry> entries, OnItemLongClickListener listener) {
        this.entries = entries;
        this.listener = listener;
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {

        TextView drug;
        TextView route;
        TextView dosage;
        TextView timestamp;

        public ViewHolder(View view) {
            super(view);

            drug = view.findViewById(R.id.textDrug);
            route = view.findViewById(R.id.textRoute);
            dosage = view.findViewById(R.id.textDosage);
            timestamp = view.findViewById(R.id.textTimestamp);
        }
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {

        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_drug_entry, parent, false);

        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {

        DrugEntry entry = entries.get(position);

        holder.drug.setText(entry.getDrug());
        holder.route.setText(entry.getRoute());
        holder.dosage.setText(entry.getDosage());

        Date date = new Date(entry.getTimestamp());

        SimpleDateFormat sdf =
                new SimpleDateFormat("yyyy-MM-dd HH:mm", Locale.getDefault());

        holder.timestamp.setText(sdf.format(date));

        holder.itemView.setOnClickListener(v -> {

            Intent intent = new Intent(v.getContext(), DrugStatisticsActivity.class);
            intent.putExtra("drug", entry.getDrug());
            v.getContext().startActivity(intent);

        });

        holder.itemView.setOnLongClickListener(v -> {
            listener.onItemLongClick(entry);
            return true;
        });
    }

    @Override
    public int getItemCount() {
        return entries.size();
    }

    public void updateEntries(List<DrugEntry> newEntries) {
        this.entries = newEntries;
        notifyDataSetChanged();
    }
}

File: app/src/main/java/com/emmikal/securedoselog/DrugAdapter.kt
package com.emmikal.securedoselog

import android.content.Intent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.emmikal.securedoselog.substances.EffectsEstimator
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

class DrugAdapter(
    private var entries: List<DrugEntry>,
    private val listener: OnItemLongClickListener
) : RecyclerView.Adapter<DrugAdapter.ViewHolder>() {

    fun interface OnItemLongClickListener {
        fun onItemLongClick(entry: DrugEntry)
    }

    class ViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        val drug: TextView = view.findViewById(R.id.textDrug)
        val route: TextView = view.findViewById(R.id.textRoute)
        val dosage: TextView = view.findViewById(R.id.textDosage)
        val timestamp: TextView = view.findViewById(R.id.textTimestamp)
        val effectsEstimate: TextView = view.findViewById(R.id.textEffectsEstimate)
        val notes: TextView = view.findViewById(R.id.textNotes)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context)
            .inflate(R.layout.item_drug_entry, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val entry = entries[position]

        holder.drug.text = entry.drug
        holder.route.text = entry.route
        holder.dosage.text = entry.dosage

        val date = Date(entry.timestamp)
        val sdf = SimpleDateFormat("yyyy-MM-dd HH:mm", Locale.getDefault())
        holder.timestamp.text = sdf.format(date)

        val estimateText = EffectsEstimator.formatEstimate(
            holder.itemView.context,
            entry
        )

        if (estimateText == null) {
            holder.effectsEstimate.visibility = View.GONE
        } else {
            holder.effectsEstimate.visibility = View.VISIBLE
            holder.effectsEstimate.text = estimateText
        }

        val note = entry.notes
        if (note.isNullOrBlank()) {
            holder.notes.visibility = View.GONE
        } else {
            holder.notes.visibility = View.VISIBLE
            holder.notes.text = note
        }

        holder.itemView.setOnClickListener { v ->
            val intent = Intent(v.context, DrugStatisticsActivity::class.java)
            intent.putExtra("drug", entry.drug)
            v.context.startActivity(intent)
        }

        holder.itemView.setOnLongClickListener {
            listener.onItemLongClick(entry)
            true
        }
    }

    override fun getItemCount(): Int = entries.size

    fun updateEntries(newEntries: List<DrugEntry>) {
        this.entries = newEntries
        notifyDataSetChanged()
    }
}